// @nx/cypress/src/generators/init/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cypressInitGenerator = void 0;
const devkit_1 = require("@nx/devkit");
const versions_1 = require("../../utils/versions");
const js_1 = require("@nx/js");
function setupE2ETargetDefaults(tree) {
    const nxJson = (0, devkit_1.readNxJson)(tree);
    if (!nxJson.namedInputs) {
        return;
    }
    // E2e targets depend on all their project's sources + production sources of dependencies
    nxJson.targetDefaults ??= {};
    const productionFileSet = !!nxJson.namedInputs?.production;
    nxJson.targetDefaults.e2e ??= {};
    nxJson.targetDefaults.e2e.cache ??= true;
    nxJson.targetDefaults.e2e.inputs ??= [
        'default',
        productionFileSet ? '^production' : '^default',
    ];
    (0, devkit_1.updateNxJson)(tree, nxJson);
}
function updateDependencies(tree) {
    (0, devkit_1.removeDependenciesFromPackageJson)(tree, ['@nx/cypress'], []);
    return (0, devkit_1.addDependenciesToPackageJson)(tree, {}, {
        ['@nx/cypress']: versions_1.nxVersion,
        cypress: versions_1.cypressVersion,
        '@types/node': versions_1.typesNodeVersion,
    });
}
function addPlugin(tree) {
    const nxJson = (0, devkit_1.readNxJson)(tree);
    nxJson.plugins ??= [];
    for (const plugin of nxJson.plugins) {
        if (typeof plugin === 'string'
            ? plugin === '@nx/cypress/plugin'
            : plugin.plugin === '@nx/cypress/plugin') {
            return;
        }
    }
    nxJson.plugins.push({
        plugin: '@nx/cypress/plugin',
        options: {
            targetName: 'e2e',
            componentTestingTargetName: 'component-test',
        },
    });
    (0, devkit_1.updateNxJson)(tree, nxJson);
}
function updateProductionFileset(tree) {
    const nxJson = (0, devkit_1.readNxJson)(tree);
    const productionFileset = nxJson.namedInputs?.production;
    if (productionFileset) {
        nxJson.namedInputs.production = Array.from(new Set([
            ...productionFileset,
            '!{projectRoot}/cypress/**/*',
            '!{projectRoot}/**/*.cy.[jt]s?(x)',
            '!{projectRoot}/cypress.config.[jt]s',
        ]));
    }
    (0, devkit_1.updateNxJson)(tree, nxJson);
}
async function cypressInitGenerator(tree, options) {
    const addPlugins = process.env.NX_PCV3 === 'true';
    updateProductionFileset(tree);
    if (!addPlugins) {
        setupE2ETargetDefaults(tree);
    }
    const tasks = [];
    tasks.push(await (0, js_1.initGenerator)(tree, {
        ...options,
        skipFormat: true,
    }));
    if (addPlugins) {
        addPlugin(tree);
    }
    if (!options.skipPackageJson) {
        tasks.push(updateDependencies(tree));
    }
    return (0, devkit_1.runTasksInSerial)(...tasks);
}
exports.cypressInitGenerator = cypressInitGenerator;
exports.default = cypressInitGenerator;
