// @nx/angular/src/generators/utils/create-ts-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTsConfig = exports.extractTsConfigBase = void 0;
const devkit_1 = require("@nx/devkit");
const js_1 = require("@nx/js");
var js_2 = require("@nx/js");
Object.defineProperty(exports, "extractTsConfigBase", { enumerable: true, get: function () { return js_2.extractTsConfigBase; } });
function createTsConfig(host, projectRoot, type, options, relativePathToRootTsConfig) {
    const json = {
        compilerOptions: {
            target: 'es2022',
            useDefineForClassFields: false,
            esModuleInterop: options.esModuleInterop ? true : undefined,
        },
        files: [],
        include: [],
        references: [
            {
                path: type === 'app' ? './tsconfig.app.json' : './tsconfig.lib.json',
            },
        ],
    };
    // inline tsconfig.base.json into the project
    if (options.rootProject) {
        json.compileOnSave = false;
        json.compilerOptions = { ...js_1.tsConfigBaseOptions, ...json.compilerOptions };
        json.exclude = ['node_modules', 'tmp'];
    }
    else {
        json.extends = relativePathToRootTsConfig;
    }
    (0, devkit_1.writeJson)(host, `${projectRoot}/tsconfig.json`, json);
}
exports.createTsConfig = createTsConfig;
