// @nx/angular/src/executors/module-federation-dev-server/lib/start-static-remotes-file-server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.startStaticRemotesFileServer = void 0;
const devkit_1 = require("@nx/devkit");
const file_server_impl_1 = require("@nx/web/src/executors/file-server/file-server.impl");
const path_1 = require("path");
const fs_1 = require("fs");
function startStaticRemotesFileServer(remotes, context, options) {
    let shouldMoveToCommonLocation = false;
    let commonOutputDirectory;
    for (const app of remotes.staticRemotes) {
        const outputPath = context.projectGraph.nodes[app].data.targets['build'].options.outputPath;
        const directoryOfOutputPath = (0, path_1.dirname)(outputPath);
        if (!commonOutputDirectory) {
            commonOutputDirectory = directoryOfOutputPath;
        }
        else if (commonOutputDirectory !== directoryOfOutputPath ||
            !outputPath.endsWith(app)) {
            shouldMoveToCommonLocation = true;
        }
    }
    if (shouldMoveToCommonLocation) {
        commonOutputDirectory = (0, path_1.join)(devkit_1.workspaceRoot, 'tmp/static-remotes');
        for (const app of remotes.staticRemotes) {
            const outputPath = context.projectGraph.nodes[app].data.targets['build'].options
                .outputPath;
            (0, fs_1.cpSync)(outputPath, (0, path_1.join)(commonOutputDirectory, app), {
                force: true,
                recursive: true,
            });
        }
    }
    const staticRemotesIter = (0, file_server_impl_1.default)({
        cors: true,
        watch: false,
        staticFilePath: commonOutputDirectory,
        parallel: false,
        spa: false,
        withDeps: false,
        host: options.host,
        port: options.staticRemotesPort,
        ssl: options.ssl,
        sslCert: options.sslCert,
        sslKey: options.sslKey,
    }, context);
    return staticRemotesIter;
}
exports.startStaticRemotesFileServer = startStaticRemotesFileServer;
