// @nx/angular/src/generators/application/lib/normalize-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeOptions = void 0;
const devkit_1 = require("@nx/devkit");
const project_name_and_root_utils_1 = require("@nx/devkit/src/generators/project-name-and-root-utils");
const eslint_1 = require("@nx/eslint");
const get_npm_scope_1 = require("@nx/js/src/utils/package-json/get-npm-scope");
const test_runners_1 = require("../../../utils/test-runners");
const project_1 = require("../../utils/project");
const version_utils_1 = require("../../utils/version-utils");
async function normalizeOptions(host, options) {
    const { projectName: appProjectName, projectRoot: appProjectRoot, projectNameAndRootFormat, } = await (0, project_name_and_root_utils_1.determineProjectNameAndRootOptions)(host, {
        name: options.name,
        projectType: 'application',
        directory: options.directory,
        projectNameAndRootFormat: options.projectNameAndRootFormat,
        rootProject: options.rootProject,
        callingGenerator: '@nx/angular:application',
    });
    options.rootProject = appProjectRoot === '.';
    options.projectNameAndRootFormat = projectNameAndRootFormat;
    const e2eProjectName = options.rootProject ? 'e2e' : `${appProjectName}-e2e`;
    const e2eProjectRoot = options.rootProject ? 'e2e' : `${appProjectRoot}-e2e`;
    const parsedTags = options.tags
        ? options.tags.split(',').map((s) => s.trim())
        : [];
    const prefix = (0, project_1.normalizeNewProjectPrefix)(options.prefix, (0, get_npm_scope_1.getNpmScope)(host), 'app');
    let bundler = options.bundler;
    if (!bundler) {
        const { major: angularMajorVersion } = (0, version_utils_1.getInstalledAngularVersionInfo)(host);
        bundler = angularMajorVersion >= 17 ? 'esbuild' : 'webpack';
    }
    // Set defaults and then overwrite with user options
    return {
        style: 'css',
        routing: true,
        inlineStyle: false,
        inlineTemplate: false,
        skipTests: options.unitTestRunner === test_runners_1.UnitTestRunner.None,
        skipFormat: false,
        unitTestRunner: test_runners_1.UnitTestRunner.Jest,
        e2eTestRunner: test_runners_1.E2eTestRunner.Cypress,
        linter: eslint_1.Linter.EsLint,
        strict: true,
        standalone: true,
        ...options,
        prefix,
        name: appProjectName,
        appProjectRoot,
        appProjectSourceRoot: `${appProjectRoot}/src`,
        e2eProjectRoot,
        e2eProjectName,
        parsedTags,
        bundler,
        outputPath: (0, devkit_1.joinPathFragments)('dist', !options.rootProject ? appProjectRoot : appProjectName),
        ssr: options.ssr ?? false,
    };
}
exports.normalizeOptions = normalizeOptions;
